File: src/experiment_client.cpp
#include <experiment/experiment_client.h>
#include <experiment/experiment_service.h>
#include <experiment/experiment_messages.h>
#include <cell_world.h>

using namespace cell_world;
using namespace tcp_messages;

namespace experiment {
    Start_experiment_response Experiment_client::start_experiment(const cell_world::World_info &world, const std::string &subject_name, int duration,
                                                   const std::string &prefix , const std::string &suffix) {
        auto parameters = Start_experiment_request();
        parameters.prefix = prefix;
        parameters.suffix = suffix;
        parameters.world = world;
        parameters.subject_name = subject_name;
        parameters.duration = duration;
        return send_request(tcp_messages::Message("start_experiment",parameters),0).get_body<Start_experiment_response>();
    }

    bool Experiment_client::start_episode(const std::string &experiment_name) {
        auto parameters = Start_episode_request();
        parameters.experiment_name = experiment_name;
        return send_request(tcp_messages::Message("start_episode",parameters),0).get_body<bool>();
    }

    bool Experiment_client::finish_episode() {
        return send_request(tcp_messages::Message("finish_episode"),0).get_body<bool>();
    }

    bool Experiment_client::finish_experiment(const std::string &experiment_name) {
        auto parameters = Finish_experiment_request();
        parameters.experiment_name = experiment_name;
        return send_request(tcp_messages::Message("finish_experiment",parameters),0).get_body<bool>();
    }

    bool Experiment_client::is_active(const std::string &experiment_name) {
        Get_experiment_response response = get_experiment(experiment_name);
        return response.remaining_time > 0;
    }

    Get_experiment_response Experiment_client::get_experiment(const std::string &experiment_name) {
        Get_experiment_request request;
        request.experiment_name = experiment_name;
        return send_request(tcp_messages::Message("get_experiment",request),0).get_body<Get_experiment_response>();
    }

    bool Experiment_client::connect(const std::string &ip) {
        auto port = Experiment_service::get_port();
        return tcp_messages::Message_client::connect(ip, port);
    }

    bool Experiment_client::set_tracking_service_ip(const std::string &ip) {
        return send_request(Message("set_tracking_service_ip", ip)).get_body<bool>();
    }

}

File: python/src/experiment_client.py
from .experiment_messages import *
from tcp_messages import MessageClient, Message
from .experiment_service import ExperimentService
from cellworld import World_info


class ExperimentClient(MessageClient):
    def __init__(self):
        MessageClient.__init__(self)
        self.router.add_route("experiment_started", self.__process_experiment_started__, StartExperimentMessage)
        self.router.add_route("episode_started", self.__process_episode_started__, int)
        self.router.add_route("episode_finished", self.__process_episode_finished__, int)
        self.router.add_route("experiment_finished", self.__process_experiment_finished__)
        self.on_experiment_start = None
        self.on_experiment_end = None
        self.on_episode_start = None
        self.on_episode_end = None

    def subscribe(self):
        return self.send_request(Message("!subscribe"), 0).body == "success"

    def __process_experiment_started__(self, parameters: StartExperimentMessage):
        if self.on_experiment_start:
            self.on_experiment_start(parameters)

    def __process_episode_started__(self, episode_number: int):
        if self.on_episode_start:
            self.on_episode_start(episode_number)

    def __process_episode_finished__(self, episode_number: int):
        if self.on_episode_end:
            self.on_episode_end(episode_number)

    def __process_experiment_finished__(self):
        if self.on_experiment_end:
            self.on_experiment_end()

    def connect(self, ip: str = "127.0.0.1"):
        MessageClient.connect(self, ip, ExperimentService.port())

    def start_experiment(self, name: str, world_configuration: str, world_implementation: str, occlusions: str, subject_name: str, duration: int):
        parameters = StartExperimentMessage(name=name, world=World_info(world_configuration, world_implementation, occlusions), subject_name=subject_name, duration=duration)
        return self.send_request(Message("start_experiment", parameters), 0).body == "success"

    def start_episode(self):
        return self.send_request(Message("start_episode"), 0).body == "success"

    def finish_episode(self):
        return self.send_request(Message("finish_episode"), 0).body == "success"

    def finish_experiment(self):
        return self.send_request(Message("finish_experiment"), 0).body == "success"




